src/public/modules/forms/helpers/response-factory.ts

```diff
@@ -0,0 +1,54 @@
+import {
+  ArrayResponse,
+  DisplayedResponseWithoutAnswer,
+  NestedResponse,
+  SingleResponse,
+} from '../../../../types/response'
+
+import {
+  ArrayAnswerResponse,
+  ErrorResponse,
+  Response,
+  SingleAnswerResponse,
+  TableResponse,
+} from './csv-response-classes'
+
+export const getResponseInstance = (
+  fieldRecordData: DisplayedResponseWithoutAnswer,
+): Response => {
+  switch (fieldRecordData.fieldType) {
+    case 'table':
+      if (isNestedResponse(fieldRecordData)) {
+        return new TableResponse(fieldRecordData)
+      }
+      break
+    case 'checkbox':
+      if (isArrayResponse(fieldRecordData)) {
+        return new ArrayAnswerResponse(fieldRecordData)
+      }
+      break
+    default:
+      if (isSingleResponse(fieldRecordData)) {
+        return new SingleAnswerResponse(fieldRecordData)
+      }
+      break
+  }
+  return new ErrorResponse(fieldRecordData)
```

Comment:
interesting approach! my concern here is that this will fail silently at the point where we actually discover the error, and then unexpectedly break something else down the line. I _think_ if you simply throw an error here, it will bubble up and get caught by `worker.on('error'` - could you help to verify this? if that's the case then I think we should rely on existing error handling mechanisms (i.e. `worker.on('error'` and `errorCount` in `submissions.client.factory`)

Reply:
I just tried and the error doesn't get caught in `worker.on('error'`. I think it will only get caught if the error occurred in the decryption worker. But I do agree this approach would be better so that its clearer that an error occurred. I'll just wrap the call to `addRecord` with a try catch block and increment `errorCount when there's an error.